ukvm/ukvm_gdb_kvm_x86_64.c

```diff
@@ -0,0 +1,460 @@
+{
+    struct kvm_guest_debug dbg = {0};
+    struct breakpoint_t *bp;
+    const uint8_t type_code[] = {
+        /* Break on instruction execution only. */
+        [GDB_BREAKPOINT_HW] = 0x0,
+        /* Break on data writes only. */
+        [GDB_WATCHPOINT_WRITE] = 0x1,
+        /* Break on data reads or writes but not instruction fetches. */
+        [GDB_WATCHPOINT_ACCESS] = 0x3
+    };
+    const uint8_t len_code[] = {
+        /*
+         * 00 — 1-byte length.
+         * 01 — 2-byte length.
+         * 10 — 8-byte length.
+         * 11 — 4-byte length.
+         */
+        [1] = 0x0, [2] = 0x1, [4] = 0x3, [8] = 0x2
+    };
+    int n = 0;
+
+    if (stepping)
+        dbg.control = KVM_GUESTDBG_ENABLE | KVM_GUESTDBG_SINGLESTEP;
+
+    if (!SLIST_EMPTY(&sw_breakpoints))
+        dbg.control |= KVM_GUESTDBG_ENABLE | KVM_GUESTDBG_USE_SW_BP;
+
+    if (!SLIST_EMPTY(&hw_breakpoints)) {
+        dbg.control |= KVM_GUESTDBG_ENABLE | KVM_GUESTDBG_USE_HW_BP;
+
+	/* Enable global breakpointing (across all threads) on the control
+         * debug register. */
+        dbg.arch.debugreg[7] = 1 << 9;
+        dbg.arch.debugreg[7] |= 1 << 10;
+        SLIST_FOREACH(bp, &hw_breakpoints, entries) {
+            dbg.arch.debugreg[n] = bp->addr;
+            /* global breakpointing */
+            dbg.arch.debugreg[7] |= (2 << (n * 2));
+            /* read/write fields */
+            dbg.arch.debugreg[7] |= (type_code[bp->type] << (16 + n*4));
+            /* Length fields */
+            dbg.arch.debugreg[7] |= ((uint32_t)len_code[bp->len] << (18 + n*4));
+            n++;
+        }
+    }
+
+    if (ioctl(hv->b->vcpufd, KVM_SET_GUEST_DEBUG, &dbg) == -1) {
+        /* The KVM_CAP_SET_GUEST_DEBUG capbility is not available. */
+        err(1, "KVM_SET_GUEST_DEBUG failed");
+        return -1;
+    }
+
+    return 0;
+}
+
+static struct breakpoint_t *bp_list_find(uint32_t type, ukvm_gpa_t addr, size_t len)
+{
+    struct breakpoint_t *bp;
+
+    if (type == GDB_BREAKPOINT_SW) {
+        SLIST_FOREACH(bp, &sw_breakpoints, entries) {
+            if (bp->addr == addr && bp->len == len)
+                return bp;
+        }
+    } else {
+        SLIST_FOREACH(bp, &hw_breakpoints, entries) {
+            if (bp->addr == addr && bp->len == len)
+                return bp;
+        }
+    }
+
+    return NULL;
+}
+
+/*
+ * Adds a new breakpoint to the list of breakpoints 
+ * Returns the found or created breakpoint, NULL if otherwise.
+ */
+static struct breakpoint_t *bp_list_insert(uint32_t type, ukvm_gpa_t addr,
+                                           size_t len)
+{
+    struct breakpoint_t *bp;
+
+    bp = bp_list_find(type, addr, len);
+    if (bp) {
+        bp->refcount++;
+        return bp;
+    }
+
+    bp = malloc(sizeof (struct breakpoint_t));
+    if (bp == NULL)
+        return NULL;
+
+    bp->addr = addr;
+    bp->type = type;
+    bp->len = len;
+    bp->refcount = 1;
+
+    if (type == GDB_BREAKPOINT_SW) {
```

Comment:
Asserting on invalid types. We should never get invalid types from GDB. And we handle all possible types.